File: incre-tests/synduce/constraints/evenlist/first_odd.f
Inductive List = elt Int | cons {Int, List};

mod2 = \x: Int. - x (* (/ x 2) 2);

is_even_pos = fix (
  \f: List -> Bool. \xs: List.
  match xs with
    elt x -> and (> x 0) (== (mod2 x) 0)
  | cons {h, t} -> and (and (> h 0) (== (mod2 h) 0)) (f t)
  end
);

spec = fix (
  \f: List -> Int. \xs: List.
  match xs with
    elt x -> if == (mod2 x) 1 then x else 0
  | cons {h, t} -> if == (mod2 h) 1 then h else f t
  end
);

target = fix (
  \f: List -> Compress List. \xs: List.
  match xs with
    elt x -> elt x
  | cons {h, t} -> cons {h, t} /*Avoid recursions*/
  end
);

main = \xs: List. if is_even_pos xs then spec (target xs) else 0;

File: ml-benchmark/synduce/constraints/evenlist/first_odd.ml
type list = Elt of int | Cons of int * list

let mod2 x = x - ((x / 2) * 2)

let rec is_even_pos xs =
  match xs with
  | Elt x -> (x > 0) && (mod2 x == 0)
  | Cons (h, t) -> ((h > 0) && (mod2 h == 0)) && is_even_pos t

let rec spec xs =
  match xs with
  | Elt x -> if mod2 x == 1 then x else 0
  | Cons (h, t) -> if mod2 h == 1 then h else spec t

val target: list -> list compress
let target xs =
  match xs with
  | Elt x -> Elt x
  | Cons (h, t) -> Cons (h, t)

let program xs =
  if is_even_pos xs then spec (target xs) else 0